api/command.go

```diff
@@ -112,55 +96,65 @@ func logoutCommand(c *Context, command *model.Command) bool {
 }
 
 func echoCommand(c *Context, command *model.Command) bool {
-
 	cmd := "/echo"
+	maxThreads := 100
 
-	if strings.Index(command.Command, cmd) == 0 {
-		parts := strings.SplitN(command.Command, " ", 3)
-
-		channelName := ""
-		if len(parts) >= 2 {
-			channelName = parts[1]
+	if !command.Suggest && strings.Index(command.Command, cmd) == 0 {
+		parameters := strings.SplitN(command.Command, " ", 2)
+		if len(parameters) != 2 || len(parameters[1]) == 0 {
+			return false
 		}
-
-		message := ""
-		if len(parts) >= 3 {
-			message = parts[2]
+		message := strings.Trim(parameters[1], " ")
+		delay := 0
+		if endMsg := strings.LastIndex(message, "\""); string(message[0]) == "\"" && endMsg > 1 {
+			if checkDelay, err := strconv.Atoi(strings.Trim(message[endMsg:], " \"")); err == nil {
+				delay = checkDelay
+			}
+			message = message[1:endMsg]
+		} else if strings.Index(message, " ") > -1 {
+			delayIdx := strings.LastIndex(message, " ")
+			delayStr := strings.Trim(message[delayIdx:], " ")
+
+			if checkDelay, err := strconv.Atoi(delayStr); err == nil {
+				delay = checkDelay
+				message = message[:delayIdx]
+			}
 		}
 
-		if result := <-Srv.Store.Channel().GetChannels(c.Session.TeamId, c.Session.UserId); result.Err != nil {
-			c.Err = result.Err
+		if delay > 10000 {
+			c.Err = model.NewAppError("echoCommand", "Delays must be under 10000 seconds", "")
 			return false
-		} else {
-			channels := result.Data.(*model.ChannelList)
+		}
 
-			for _, v := range channels.Channels {
-				if v.Type == model.CHANNEL_DIRECT {
-					continue
-				}
+		if echoSem == nil {
+			// We want one additional thread allowed so we never reach channel lockup
+			echoSem = make(chan bool, maxThreads+1)
```

Comment:
I was thinking about doing that at first. Ultimately, it probably would be the same, but since I'm using go-routines I figured it was safer to use a concurrency primitive and not risk a variable being controlled by multiple threads with no check. I'm 0/5 honestly, especially since there's no real performance loss to using a channel.